design/automatic-network-configuration.md

```diff
@@ -0,0 +1,485 @@
+When this CR is created, the operation of configuring network will be triggered,
+and when this CR is deleted, the operation of clearing network configuration will
+be triggered.
+This CR cannot be modified. If user wants to change the network configuration
+applied on a port, it is needed to clear the configuration first and then apply
+new configuration. So modification is done by deleting old `NetworkBinding` CR
+and creating new one.
+
+```yaml
+metaData:
+  name:
+  ownerRef:
+  finalizers:
+spec:
+  networkConfigurationRef:
+    name:
+    namespace:
+  ports:
+  - portID:
+    deviceRef:
+      name:
+      kind:
+      namespace:
+  - portID:
+    deviceRef:
+      name:
+      kind:
+      namespace:
+status:
+  state:
+```
+
+#### NetworkBinding Controller
+
+Add the environment variable `ADMIN_NAMESPACE`, the NetworkBinding Controller
+judges whether the user has the authority to configure the port according to the
+content of `Device CR annotation`.
+
+* User creates `NetworkBinding` CR
+  * Instantiate the specific `DeviceManager` according to the `deviceRef`
+    information in the `.spec.ports` field, and then call the
+    `ConfigureNetwork()` method to modify the corresponding `Device` CR.
+
+* User deletes `NetworkBinding` CR
+  * Instantiate the specific `DeviceManager` according to the `deviceRef`
+    information in the `.spec.ports` field, and then call the
+    `DeConfigureNetwork()` method to modify the corresponding `Device` CR.
+
+#### Switch
+
+```yaml
+kind: Switch
+metaData:
+  name: switch0
+  finalizers:
+  annotations:
+    adminOnlyPorts: "0/22,0/23,3/34-3/45"
+    reservedPorts: "0/1,0/2"
+spec:
+  os: fos
+  ip: 192.168.0.1
+  mac: ff:ff:ff:ff:ff:ff
+  secret:
+  # Indicates the switch at the other end of the peer link with this switch.
+  peerLinkWith:
+  - name: switch1
+    namespace: def
+status:
+  ports:
+  - portID:
+    # Current network configuration applied to this port
+    networkConfiguration:
+    # Reference refer to desired networkConfiguration CR
+    networkConfigurationRef:
+      name:
+      namespace:
+    lagWith:
+    - portID:
+      deviceRef:
+        name:
+        kind:
+        namespace:
+    state:
+```
+
+#### Switch Controller
+
+* Administrator creates `Switch` CR
+  * The Switch Controller checks whether the data entered by the user is legal,
+    and adds `finalizers`, then uses `Ansible` to obtain all port information
+    of the switch, and initializes all ports in `.status.ports`.
+
+* Administrator deletes `Switch` CR
+  * The Switch Controller detects whether the number of configured ports is 0.
+    If it is not 0, it is not allowed to delete the CR, otherwise delete the
+    `finalizers` and delete the CR. The controller determines whether a port is
+    configured by checking the port's `.networkConfiguration`. If this field is not
+    nil, the port is configured.
+
+* The NetworkBinding Controller modify `.status.ports`
```

Comment:
I would say that the status of a resource should only be modified by its own controller

Reply:
updated.